### PR TITLE
ppx_optcomp injection + removal of consensus environment configuration

### PR DESCRIPTION
**Pull request by nholland94**
This PR contains 2 major changes in order to address the issues we have been seeing lately with consensus configuration. Firstly, the consensus configuration has been made explicitly compile time only, preventing us from having snark failures due to environment variable mismatches between build time and run time. Secondly, the consensus mechanism was defunctored and raw implementations were hidden in order to enforce that all code generators see the same consensus mechanism as the compiled code. This helps protect against some errors that were found last week.

This PR may seem daunting, but the changes can really be broken up into 3 parts:

1) `ppx_optcomp` was forked and modified to support injecting configuration values into ocaml structures. The diff for these changes can be viewed here: https://github.com/CodaProtocol/ppx_optcomp/compare/159caef808dfc7a069cff627723109391d63cfa7...master
2) `Consensus` was refactored to remove functors and hide the individual implementations properly. References were then updated, which was mostly automated, with the exception of one bug that was found by doing this which I have fixed.
3) The configuration files have been restructured to allow sharing of common configuration.

In particular, I would like to ask @jkrauska to review the configuration changes (particularly to makes sure that `testnet_public` and the like are in the proper state you want), and @cmr to double check to see if the `fake_hash` profile is how he wants it to be.

### CODE REVIEW

**Comment by cmr** [comment]
`fake_hash` looks fine.

**Comment by nholland94** [comment]
@jkrauska I'm getting a failure on `make deb` now as I removed the `CODA_CONSENSUS` env, which is used by `rebuild-deb.sh` to configure the package name. With the new build system the way it is now, can I change this to just use `DUNE_PROFILE` and name the package by the build profile?

**Comment by jkrauska** [comment]
> @jkrauska I'm getting a failure on `make deb` now as I removed the `CODA_CONSENSUS` env, which is used by `rebuild-deb.sh` to configure the package name. With the new build system the way it is now, can I change this to just use `DUNE_PROFILE` and name the package by the build profile?

I found DUNE_PROFILE too verbose.  That's why I created CODA_CONSENSUS.

As a hack, sure, use DUNE_PROFILE..  or put it back in place?

File: Makefile [hunk 1]
```diff
@@ -67,7 +67,7 @@ dht: kademlia
 
 build: git_hooks
 	$(info Starting Build)
-	ulimit -s 65532 && (ulimit -n 10240 || true) && (ulimit -u 2128 || true) && cd src && $(WRAPSRC) env CODA_COMMIT_SHA1=$(GITLONGHASH) dune build --profile=$(DUNE_PROFILE)
```
Comment (bkase) [inline Makefile]: I think the `ulimit -u` is required for macos builds, but if CI passes, than I'm down with it (the mac default is 1418)

(this may be because we were accidentally building on `-j36` on CI before)

File: Makefile [hunk 2]
```diff
@@ -235,7 +235,7 @@ test-stakes:
 
 test-withsnark: SHELL := /bin/bash
 test-withsnark:
-	source scripts/test_all.sh ; cd src; CODA_BLOCK_DURATION=30000 WITH_SNARKS=true DUNE_PROFILE=test_snark run_integration_test full-test
+	source scripts/test_all.sh ; cd src; CODA_BLOCK_DURATION=30000 WITH_SNARKS=true DUNE_PROFILE=test_posig run_integration_test full-test
```
Comment (bkase) [inline Makefile]: `CODA_BLOCK_DURATION` is gone right?

File: src/config/testnet_postake.mlh [hunk 3]
```diff
@@ -1,10 +1,11 @@
+[%%import "config/coinbase/standard.mlh"]
+[%%import "config/consensus/postake_short.mlh"]
+[%%import "config/debug_level/logs.mlh"]
+
 [%%define genesis_ledger "testnet_postake"]
+[%%define genesis_state_timestamp "2019-01-30 12:00:00-08:00"]
+[%%define block_window_duration 20000]
```
Comment (bkase) [inline src/config/testnet_postake.mlh]: I think we want 60000 with snarks on for now

File: src/dune [hunk 4]
```diff
@@ -1,5 +1,25 @@
-(rule
-  (targets config.mlh)
-  (mode fallback)
-  (deps config/%{profile}.mlh)
-  (action (copy "%{deps}" config.mlh)))
+(* -*- tuareg -*- *)
```
Comment (bkase) [inline src/dune]: does this make it harder to make merlin happy with our source tree?

File: src/lib/consensus/proof_of_stake.ml [hunk 5]
```diff
@@ -41,1433 +50,1381 @@ let uint32_of_int64 x = x |> Int64.to_int64 |> UInt32.of_int64
...
+end
 
-    let prover_state {stake_proof; _} = stake_proof
-  end
+let genesis_ledger_total_currency =
+  Coda_base.Ledger.to_list Genesis_ledger.t
+  |> List.fold_left ~init:Balance.zero ~f:(fun sum account ->
+         Balance.add_amount sum
+           (Balance.to_amount @@ Coda_base.Account.balance account)
+         |> Option.value_exn ?here:None ?error:None
```
Comment (bkase) [inline src/lib/consensus/proof_of_stake.ml]: why did you add these optional params?

File: Makefile (hunk 2 above)
Comment (nholland94) [reply Makefile]: yes, let me fix that

File: Makefile (hunk 1 above)
Comment (nholland94) [reply Makefile]: I think this was an accident. I will add it back in for now, even though CI works. Unless you think we should limit our calls to ulimit.

File: src/config/testnet_postake.mlh (hunk 3 above)
Comment (nholland94) [reply src/config/testnet_postake.mlh]: Yes, good call.

File: src/dune (hunk 4 above)
Comment (nholland94) [reply src/dune]: Not sure if I understand what you mean. This will not effect merlin in any way.

File: src/lib/consensus/proof_of_stake.ml (hunk 5 above)
Comment (nholland94) [reply src/lib/consensus/proof_of_stake.ml]: For some reason, it started generating a type error. Really uncertain as to why though. I think maybe because I converted the jbuild file into a dune file?
